File: src/bench/test/scala/scife/enumeration/benchmarks/iterators/java.scala
package scife.enumeration.benchmarks.iterators

import scife.enumeration.benchmarks.java._

// Generates Java expressions and statements with the intention
// of testing APIs.  Assumes there are existing classes around.
// Only interested in creating new objects and calling methods.
// Does not handle generics.

object JavaGeneration {
  type Env = Map[Variable, Type]

  val MIN_NUM_STATEMENTS = 1
  val MAX_NUM_STATEMENTS = 1

  val objectClass = ClassName("Object")

  val DEFAULT_SPEC: List[ClassSpec] =
    List(ClassSpec(objectClass,
		   Set(),
		   List(ConstructorSpec(List())),
		   List()),
	 ClassSpec(ClassName("CharSequence"),
		   Set(objectClass),
		   List(),
		   List(MethodSpec(MethodName("length"),
				   List(),
				   Some(IntType)))),
	 ClassSpec(ClassName("String"),
		   Set(objectClass, ClassName("CharSequence")),
		   List(ConstructorSpec(List())),
		   List(MethodSpec(MethodName("concat"),
				   List(ClassType(ClassName("String"))),
				   Some(ClassType(ClassName("String")))))),
	 ClassSpec(ClassName("java.util.ArrayList"),
		   Set(objectClass),
		   List(ConstructorSpec(List()),
			ConstructorSpec(List(IntType))),
		   List(MethodSpec(MethodName("size"),
				   List(),
				   Some(IntType)),
			MethodSpec(MethodName("add"),
				   List(ClassType(objectClass)),
				   Some(BooleanType)),
			MethodSpec(MethodName("remove"),
				   List(ClassType(objectClass)),
				   Some(BooleanType)))))

  def disjuncts[A](iterators: Iterator[A]*): Iterator[A] = {
    iterators.reduceRight(_ ++ _)
  } // disjuncts

  def main(args: Array[String]) {
    if (args.length != 1) {
      println("Needs a size bound")
    } else {
      val bound = args(0).toInt
      val generator = apply(1000)
      generator.makeTest(bound).foreach(test => println(test.prettyString))
    }
  } // main

  def apply(k: Int): JavaGeneration = {
    new JavaGeneration(DEFAULT_SPEC) {
      def makeCache[K, V]: IteratorCache[K, V] = new RealCacheForceSize(k)
    }
  } // apply
}
import JavaGeneration._

abstract class JavaGeneration(val spec: List[ClassSpec]) {
  // for a given class, these are the parents
  val hierarchy: Map[ClassName, Set[ClassName]] =
    spec.map(s => (s.name, s.extendsImplements)).toMap

  val allParentsCache: IteratorCache[ClassName, ClassName] = makeCache
  val makeExpressionsOfTypesCache: IteratorCache[(Int, Env, List[Type]), List[Expression]] = makeCache
  val makeCallCache: IteratorCache[(Int, Env, Boolean), (Call, Option[Type])] = makeCache
  val makeExpressionCache: IteratorCache[(Int, Env), (Expression, Type)] = makeCache
  val makeStatementCache: IteratorCache[(Int, Env, Int), (Statement, Env)] = makeCache
  val makeStatementsCache: IteratorCache[(Int, Env, Int), List[Statement]] = makeCache
  val makeTestCache: IteratorCache[Int, JavaTest] = makeCache

  def allCaches = (allParentsCache,
		   makeExpressionsOfTypesCache,
		   makeCallCache,
		   makeExpressionCache,
		   makeStatementCache,
		   makeStatementsCache,
		   makeTestCache)

  def makeCache[K, V]: IteratorCache[K, V]

  def allParents(of: ClassName): Iterator[ClassName] = {
    allParentsCache.tryCache(of, {
      hierarchy.get(of) match {
	case Some(directParents) => {
	  disjuncts(directParents.iterator,
		    directParents.iterator.flatMap(allParents))
	}
	case None => Iterator()
      }
    })
  } // sameAndAllParents

  def parentChild(parent: ClassName, child: ClassName): Boolean = {
    hierarchy.get(child) match {
      case Some(directParents) => {
	if (directParents.contains(parent)) {
	  true
	} else {
	  directParents.exists(directParent => parentChild(parent, directParent))
	}
      }
      case None => false
    }
  } // parentChild

  def typesCompatibleParentChild(parent: Type, child: Type): Boolean = {
    (parent, child) match {
      case (`child`, `parent`) => true
      case (ClassType(parentName), ClassType(childName)) => parentChild(parentName, childName)
      case _ => false
    }
  } // typesCompatibleParentChild

  def makeExpressionsOfTypes(size: Int, env: Env, types: List[Type]): Iterator[List[Expression]] = {
    makeExpressionsOfTypesCache.tryCache((size, env, types), {
      types match {
	case Nil => Iterator(List())
	case head :: tail => {
	  for {
	    (expression, expressionType) <- makeExpression(size, env)
	    // given expression can be a subtype of what we expect
	    if typesCompatibleParentChild(head, expressionType)
	    rest <- makeExpressionsOfTypes(size, env, tail)
	  } yield expression :: rest
	}
      }
    })
  } // makeExpressionsOfTypes

  def makeCall(size: Int, env: Env, voidOk: Boolean): Iterator[(Call, Option[Type])] = {
    makeCallCache.tryCache((size, env, voidOk), {
      for {
	(base, ClassType(myClassName)) <- makeExpression(size, env)
	methodClassName <- disjuncts(Iterator(myClassName), allParents(myClassName))
	MethodSpec(methodName, paramTypes, returnType) <- spec.find(_.name == methodClassName).get.methods.iterator
	if returnType.isDefined || voidOk
	params <- makeExpressionsOfTypes(size, env, paramTypes)
      } yield (Call(base, methodName, params), returnType)
    })
  } // makeCall
  
  def makeExpression(size: Int, env: Env): Iterator[(Expression, Type)] = {
    makeExpressionCache.tryCache((size, env), {
      val baseCases: Iterator[(Expression, Type)] =
	disjuncts(Iterator((IntLiteral(0), IntType)),
		  Iterator((BooleanLiteral(false), BooleanType)),
		  Iterator((DoubleLiteral(0.0), DoubleType)),
		  Iterator((StringLiteral("foo"), ClassType(ClassName("String")))),
		  env.iterator.map(pair => (VariableExpression(pair._1), pair._2)))

      if (size <= 0) {
	baseCases
      } else {
	disjuncts(baseCases,
		  // new
		  for {
		    ClassSpec(className, _, constructors, _) <- spec.iterator
		    ConstructorSpec(paramTypes) <- constructors.iterator
		    params <- makeExpressionsOfTypes(size - 1, env, paramTypes)
		  } yield (NewExpression(className, params), ClassType(className)),
		  // call - non-void
		  for {
		    (call, Some(returnType)) <- makeCall(size - 1, env, false)
		  } yield (CallExpression(call), returnType))
      }
    })
  } // makeExpression
				 
  def makeStatement(size: Int, env: Env, statementNum: Int): Iterator[(Statement, Env)] = {
    makeStatementCache.tryCache((size, env, statementNum), {
      disjuncts(// variable declaration
                for {
		  (initializer, initializerType) <- makeExpression(size, env)
		} yield {
		  val variable = Variable("x" + statementNum)
		  (VariableDeclaration(initializerType, variable, initializer), env + (variable -> initializerType))
		},
                // call - possibly void
                for {
		  (call, _) <- makeCall(size, env, true)
		} yield (CallStatement(call), env))
    })
  } // makeStatement

  def makeStatements(size: Int, env: Env, numStatements: Int): Iterator[List[Statement]] = {
    makeStatementsCache.tryCache((size, env, numStatements), {
      if (numStatements < 0) {
	Iterator()
      } else if (numStatements == 0) {
	Iterator(List())
      } else {
	assert(numStatements > 0)
	for {
	  (statement, newEnv) <- makeStatement(size, env, numStatements)
	  rest <- makeStatements(size, newEnv, numStatements - 1)
	} yield statement :: rest
      }
    })
  } // makeStatements

  def makeTest(size: Int): Iterator[JavaTest] = {
    makeTestCache.tryCache(size, {
      for {
	numStatements <- MIN_NUM_STATEMENTS.to(MAX_NUM_STATEMENTS).iterator
	statements <- makeStatements(size, Map(), numStatements)
      } yield JavaTest(statements.toSeq)
    })
  } // makeTest
} // JavaGeneration

File: clp/java.pl
module(java,
       [java_main/2],
       [java_test]).

datadef(pair, [A, B], [pair(A, B)]).

datadef(map, [K, V], [map(list(pair(K, V)),
                          relation([K, K]),
                          relation([K, K]))]).

clausedef(mapMember, [K, V], [K,           % key to look for
                              V,           % value bound for that key
                              map(K, V)]). % the map to look in
mapMember(K, V, map(Pairs, _, _)) :-
    member(pair(K, V), Pairs).

clausedef(mapAdd, [K, V], [map(K, V),   % input map
                           K,           % key to add
                           V,           % value to add for the key
                           map(K, V)]). % output map
mapAdd(map(InPairs, R1, R2), K, V, map(OutPairs, R1, R2)) :-
    mapAddInternal(InPairs, R1, R2, K, V, OutPairs).

clausedef(mapAddInternal, [K, V], [list(pair(K, V)),   % input map
                                   relation([K, K]),   % used to determine if two keys are the same
                                   relation([K, K]),   % used to determine if two keys are different
                                   K,                  % key to add
                                   V,                  % value to add for the key
                                   list(pair(K, V))]). % output map
mapAddInternal([], _, _, K, V, [pair(K, V)]).
mapAddInternal([pair(K1, _)|Rest], KeysSame, _, K2, V, [pair(K2, V)|Rest]) :-
    call(KeysSame, K1, K2).
mapAddInternal([Pair|InRest], KeysSame, KeysDifferent, K1, V, [Pair|OutRest]) :-
    Pair = pair(K2, _),
    call(KeysDifferent, K1, K2),
    mapAddInternal(InRest, KeysSame, KeysDifferent, K1, V, OutRest).

clausedef(makeEqualsMap, [K, V], [map(K, V)]).
makeEqualsMap(map([],
                  lambda([K1, K2], K1 == K2),
                  lambda([K1, K2], K1 \== K2))).


datadef(env, [], [env(map(variable, type))]).

datadef(variable, [], [variable(atom)]).
datadef(method_name, [], [method_name(atom)]).
datadef(class_name, [], [class_name(atom)]).

datadef(type,
        [],
        [int_type,
         boolean_type,
         double_type,
         class_type(class_name)]).

datadef(call,
        [],
        [call(expression,
              method_name,
              list(expression))]).

datadef(boolean,
        [],
        [yes, no]).

datadef(expression,
        [],
        [variable_expression(variable),
         int_literal(int),
         boolean_literal(boolean),
         double_literal(int),
         string_literal(atom),
         new_expression(class_name, list(expression)),
         call_expression(call)]).

datadef(statement,
        [],
        [variable_declaration(type, variable, expression),
         call_statement(call)]).

datadef(java_test,
        [],
        [java_test(list(statement))]).

datadef(class_spec,
        [],
        [class_spec(class_name,
                    list(class_name),
                    list(constructor_spec),
                    list(method_spec))]).

datadef(constructor_spec,
        [],
        [constructor_spec(list(type))]).

datadef(option,
        [A],
        [some(A),
         none]).

datadef(method_spec,
        [],
        [method_spec(method_name,
                     list(type),
                     option(type))]).

clausedef(defaultSpec, [], [list(class_spec)]).
defaultSpec([class_spec(class_name('Object'),
                        [],
                        [constructor_spec([])],
                        []),
             class_spec(class_name('CharSequence'),
                        [class_name('Object')],
                        [],
                        [method_spec(method_name('length'),
                                     [],
                                     some(int_type))]),
             class_spec(class_name('String'),
                        [class_name('Object'), class_name('CharSequence')],
                        [constructor_spec([])],
                        [method_spec(method_name('concat'),
                                     [class_type(class_name('String'))],
                                     some(class_type(class_name('String'))))]),
             class_spec(class_name('java.util.ArrayList'),
                        [class_name('Object')],
                        [constructor_spec([]),
                         constructor_spec([int_type])],
                        [method_spec(method_name('size'),
                                     [],
                                     some(int_type)),
                         method_spec(method_name('add'),
                                     [class_type(class_name('Object'))],
                                     some(boolean_type)),
                         method_spec(method_name('remove'),
                                     [class_type(class_name('Object'))],
                                     some(boolean_type))])]).

% gets all the immediate parents of the given class
clausedef(hierarchyGet, [], [list(class_spec),
                             class_name,
                             list(class_name)]).
hierarchyGet(Spec, ClassName, DirectParents) :-
    once(lambda([], member(class_spec(ClassName, DirectParents, _, _), Spec))).

clausedef(allParents, [], [list(class_spec),
                           class_name,
                           class_name]).
allParents(Spec, Of, Result) :-
    hierarchyGet(Spec, Of, DirectParents),
    member(DirectParent, DirectParents),
    (Result = DirectParent;
     allParents(Spec, DirectParent, Result)).

clausedef(once, [], [relation([])]).
once(Call) :-
    call(Call),
    !.
    
% succeeds if this is a parent/child relationship
clausedef(parentChild, [], [list(class_spec),
                            class_name,
                            class_name]).
parentChild(Spec, Parent, Child) :-
    hierarchyGet(Spec, Child, DirectParents),
    member(Parent, DirectParents),
    !.
parentChild(Spec, Parent, Child) :-
    hierarchyGet(Spec, Child, DirectParents),
    member(DirectParent, DirectParents),
    parentChild(Spec, Parent, DirectParent),
    !.

% succeeds if true
clausedef(typesCompatibleParentChild, [], [list(class_spec),
                                           type,
                                           type]).
typesCompatibleParentChild(_, Type, Type) :- !.
typesCompatibleParentChild(Spec, class_type(ParentName), class_type(ChildName)) :-
    parentChild(Spec, ParentName, ChildName),
    !.

clausedef(makeExpressionsOfTypes, [], [list(class_spec),
                                       int,
                                       env,
                                       list(type),
                                       list(expression)]).
makeExpressionsOfTypes(_, _, _, [], []).
makeExpressionsOfTypes(Spec, Size, Env, [Type|Types], [Expression|Expressions]) :-
    makeExpression(Spec, Size, Env, Expression, ActualType),
    typesCompatibleParentChild(Spec, Type, ActualType),
    makeExpressionsOfTypes(Spec, Size, Env, Types, Expressions).

clausedef(findCompatibleMethod, [], [list(class_spec),
                                     class_name,
                                     method_spec]).
findCompatibleMethod(Spec, ClassName, MethodSpec) :-
    member(class_spec(ClassName, _, _, MethodSpecs), Spec),
    member(MethodSpec, MethodSpecs).

clausedef(returnTypeOk, [], [option(type), boolean]).
returnTypeOk(some(_), _) :- !.
returnTypeOk(_, yes).

clausedef(makeCall, [], [list(class_spec),
                         int,
                         env,
                         boolean,
                         call,
                         option(type)]).
makeCall(Spec, Size, Env, VoidOk, call(Base, MethodName, Params), ReturnType) :-
    makeExpression(Spec, Size, Env, Base, ExpressionType),
    ExpressionType = class_type(MyClassName), % intentionally mirroring Scala
    (MethodClassName = MyClassName;
     allParents(Spec, MyClassName, MethodClassName)),
    findCompatibleMethod(Spec, MethodClassName, method_spec(MethodName, ParamTypes, ReturnType)),
    returnTypeOk(ReturnType, VoidOk),
    makeExpressionsOfTypes(Spec, Size, Env, ParamTypes, Params).

clausedef(decBound, [], [int, int]).
decBound(In, Out) :-
    In > 0,
    Out is In - 1.

clausedef(makeExpression, [], [list(class_spec),
                               int,
                               env,
                               expression,
                               type]).
makeExpression(_, _, _, int_literal(0), int_type).
makeExpression(_, _, _, boolean_literal(no), boolean_type).
makeExpression(_, _, _, double_literal(0), double_type).
makeExpression(_, _, _, string_literal('foo'), class_type(class_name('String'))).
makeExpression(_, _, env(Map), variable_expression(Variable), Type) :-
    mapMember(Variable, Type, Map).
makeExpression(Spec, Size1, Env, new_expression(ClassName, Params), class_type(ClassName)) :-
    decBound(Size1, Size2),
    member(class_spec(ClassName, _, Constructors, _), Spec),
    member(constructor_spec(ParamTypes), Constructors),
    makeExpressionsOfTypes(Spec, Size2, Env, ParamTypes, Params).
makeExpression(Spec, Size1, Env, call_expression(Call), ReturnType) :-
    decBound(Size1, Size2),
    makeCall(Spec, Size2, Env, no, Call, RawReturnType),
    RawReturnType = some(ReturnType).

clausedef(yolo_UNSAFE_atom_concat_number, [], [atom, int, atom]).
yolo_UNSAFE_atom_concat_number(Atom, Int, Result) :-
    number_atom(Int, IntAtom),
    atom_concat(Atom, IntAtom, Result).

clausedef(makeStatement, [], [list(class_spec),
                              int,
                              env,
                              int,
                              statement,
                              env]).
makeStatement(Spec, Size, Env, StatementNum,
              variable_declaration(InitializerType, Variable, Initializer),
              env(NewMap)) :-
    makeExpression(Spec, Size, Env, Initializer, InitializerType),
    yolo_UNSAFE_atom_concat_number('x', StatementNum, VariableName),
    Variable = variable(VariableName),
    Env = env(Map),
    mapAdd(Map, Variable, InitializerType, NewMap).
makeStatement(Spec, Size, Env, _,
              call_statement(Call),
              Env) :-
    makeCall(Spec, Size, Env, yes, Call, _).

clausedef(makeStatements, [], [list(class_spec),
                               int, % size
                               env,
                               int, % numStatements
                               list(statement)]).
makeStatements(_, _, _, 0, []).
makeStatements(Spec, Size, Env, NumStatements, [Statement|Statements]) :-
    NumStatements > 0,
    makeStatement(Spec, Size, Env, NumStatements, Statement, NewEnv),
    NewNumStatements is NumStatements - 1,
    makeStatements(Spec, Size, NewEnv, NewNumStatements, Statements).

clausedef(makeTest, [], [int, java_test]).
makeTest(Bound, java_test(Statements)) :-
    defaultSpec(Spec),
    makeEqualsMap(EmptyMap),
    makeStatements(Spec, Bound, env(EmptyMap), 1, Statements).

clausedef(java_main, [], [int, java_test]).
java_main(Size, Test) :-
    makeTest(Size, Test).
